misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt

```diff
@@ -0,0 +1,381 @@
+package misk.jdbc
+
+import com.squareup.moshi.Moshi
+import misk.moshi.adapter
+import misk.okio.split
+import mu.KotlinLogging
+import net.ttddyy.dsproxy.ExecutionInfo
+import net.ttddyy.dsproxy.QueryInfo
+import net.ttddyy.dsproxy.listener.MethodExecutionContext
+import net.ttddyy.dsproxy.listener.MethodExecutionListener
+import net.ttddyy.dsproxy.listener.QueryExecutionListener
+import net.ttddyy.dsproxy.proxy.ProxyConfig
+import net.ttddyy.dsproxy.support.ProxyDataSource
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okio.BufferedSource
+import okio.ByteString.Companion.encodeUtf8
+import java.io.EOFException
+import java.sql.Connection
+import javax.inject.Singleton
+import javax.sql.DataSource
+
+internal data class Instruction(
+  val Opcode: String,
+  val Input: Instruction?
+) {
+  val isScatter: Boolean
+    get() {
+      return Opcode == "SelectScatter" || Input?.isScatter ?: false
+    }
+}
+
+internal data class QueryPlan(
+  val Original: String,
+  val Instructions: Instruction,
+  val ExecCount: Int
+) {
+  val isScatter: Boolean
+    get() = Instructions.isScatter
+}
+
+internal data class Variables(
+  /** Queries processed by plan type */
+  val QueriesProcessed: Map<String, Int>
+)
+
+open class ExtendedQueryExectionListener : QueryExecutionListener, MethodExecutionListener {
+  override fun beforeMethod(executionContext: MethodExecutionContext) {
+    if (isStartTransaction(executionContext)) {
+      beforeStartTransaction()
+    }
+    if (isRollbackTransaction(executionContext)) {
+      doBeforeRollback()
+    }
+    if (isCommitTransaction(executionContext)) {
+      doBeforeCommit()
+    }
+  }
+
+  override fun afterMethod(executionContext: MethodExecutionContext) {
+    if (isStartTransaction(executionContext)) {
+      afterStartTransaction()
+    }
+    if (isRollbackTransaction(executionContext)) {
+      doAfterRollback()
+    }
+    if (isCommitTransaction(executionContext)) {
+      doAfterCommit()
+    }
+  }
+
+  private fun isStartTransaction(executionContext: MethodExecutionContext) =
+      executionContext.method.name == "setAutoCommit" && executionContext.methodArgs[0] == false
+
+  private fun isRollbackTransaction(executionContext: MethodExecutionContext) =
+      executionContext.method.name == "rollback"
+
+  private fun isCommitTransaction(executionContext: MethodExecutionContext) =
+      executionContext.method.name == "commit"
+
+  final override fun beforeQuery(execInfo: ExecutionInfo?, queryInfoList: MutableList<QueryInfo>?) {
+    if (queryInfoList == null) return;
```

Comment:
simpler to enforce non-null and make the callers not call it? As-is it’s weird to have a method that accepts null but just bails if it’s passed in

Reply:
I have no control over the callers though.